src/http/StatusLine.cc

```diff
@@ -51,9 +51,6 @@ Http::StatusLine::packInto(Packable * p) const
     auto packedReason = reason();
 
     if (packedStatus == Http::scNone) {
-        static unsigned int reports = 0;
-        if (++reports <= 100)
-            debugs(57, DBG_IMPORTANT, "BUG: Generated response lacks status code");
```

Comment:
The removed code protected Squid from packing responses with invalid status code. AFAICT, this PR removes that protection. The PR also adds similar protection for one specific use case (PURGE transactions), but that is not an equivalent replacement AFAICT.

Can we keep the original universal protection without triggering BUG messages for responses received with zero status codes? Do we have to pack those responses "as is", including the invalid zero status code, for some reason?

Reply:
> The removed code protected Squid from packing responses with invalid status code. 

As I see it, the protection still remains below, meaning that an invalid scNone will be replaced with scInternalServerError in the packed message. The level-1 logging (which I moved from here) did not protect from this, right?

> The PR also adds similar protection for one specific use case (PURGE transactions)

My understanding is based on the SQUID-522 [discussion](https://measurementfactory.atlassian.net/browse/SQUID-522?focusedCommentId=42592&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-42592): 

```
If the code reaches response generation phase with purgeStatus still unset,
then something went wrong and we should probably generate an internal error page
```
Note that the discussed case is specific and about purgeStatus, I could not find any justification there to apply this protection to all possible cases.

> Can we keep the original universal protection without triggering BUG messages for responses received with zero status codes?

That is what I tried to do here: I made the BUG message specific to PURGEs (as probably PR522 intention was) and left the protection unchanged.

>  Do we have to pack those responses "as is", including the invalid zero status code, for some reason?

I think no (in a case of received response), but we should probably use [502 Bad Gateway](https://tools.ietf.org/html/rfc7231#page-63), which better suits for an invalid status code response in this case, which is:
```
The 502 (Bad Gateway) status code indicates that the server, while
acting as a gateway or proxy, received an invalid response from an
inbound server it accessed while attempting to fulfill the request.
```

Reply:
> the protection still remains below, meaning that an invalid scNone will be replaced with scInternalServerError in the packed message

I regret using the vague/unqualified term "protection". It muddied the discussion further and probably even confused me later in the change request.

There are several layers here:

1. Yes, the existing protection from packing zero status codes remains in effect. We should keep it there (until a better solution is found). No related changes are expected in this PR.
2. The code bringing admin/developers attention to the problem was moved, reducing its coverage. We should move it back, restoring its more general location and wider scope. Once the original location is restored, this  PR will show no changes in that area either.
3. There was and still is no protection from before-pack() code misusing zero status code. We may not find a good general solution for this problem, but this PR should fix the newly discovered (or alleged) "received zero status code" use case. These changes are missing from this PR.
4. There are also purging code changes, but they have a dedicated change request so I am not going to discuss them here.
> Note that the discussed case is specific and about purgeStatus,

Yes, it was about specific problem we were dealing with at the time. I think those two enumerated [principles](https://measurementfactory.atlassian.net/browse/SQUID-522?focusedCommentId=42592&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-42592) are still valid as far as purging code is concerned. I hope that our fix for the purging code works fine, but please correct me if you think that (now official) purging code is buggy.

> I could not find any justification there to apply this protection to all possible cases.

I think the warnings in pack() are useful. They can help us catch more bugs like those fixed in SQUID-522. The _fix_ for the SQUID-522 bug is specific to that bug, but the warnings are more general. I am glad those warnings exposed the "Squid uses received zero status code" problem alleged in bug 5057, for example.

> Alex: Do we have to pack those responses "as is", including the invalid zero status code, for some reason?

Sorry, my question was based on the wrong premise: We do not actually pack those responses as is. I withdraw this question.